src/miral/application_selector.cpp

```diff
@@ -0,0 +1,223 @@
+    {
+        selected = nullptr;
+    }
+}
+
+void ApplicationSelector::advise_delete_app(Application const& application)
+{
+    auto it = std::find(focus_list.begin(), focus_list.end(), application);
+    if (it == focus_list.end())
+    {
+        mir::log_warning("ApplicationSelector::advise_delete_app could not delete the app.");
+        return;
+    }
+
+    // If we delete the selected application, we will try to select the next available one.
+    if (application == selected)
+    {
+        auto original_it = it;
+        do {
+            it++;
+            if (it == focus_list.end())
+                it = focus_list.begin();
+
+            if (it == original_it)
+            {
+                break;
+            }
+        } while (!tools.can_select_application(*it, active_window));
+
+        if (it != original_it)
+            selected = *it;
+    }
+
+    focus_list.erase(it);
+}
+
+auto ApplicationSelector::next(bool reverse) -> Application
+{
+    if (focus_list.empty())
+    {
+        return nullptr;
+    }
+
+    if (!is_active())
+    {
+        originally_selected = focus_list.front();
+        selected = originally_selected;
+    }
+
+    // Attempt to focus the next application after the selected application.
+    auto it = std::find(focus_list.begin(), focus_list.end(), selected);
+    auto start_it = it;
+
+    Window next_window;
+    do {
+        if (reverse)
+        {
+            if (it == focus_list.begin())
+            {
+                it = focus_list.end() - 1;
+            }
+            else
+            {
+                it--;
+            }
+        }
+        else
+        {
+            it++;
+            if (it == focus_list.end())
+            {
+                it = focus_list.begin();
+            }
+        }
+
+        // We made it all the way through the list but failed to find anything.
+        if (it == start_it)
+            break;
+    } while (!tools.can_select_application(*it, next_window));
+
+    // Swap the tree order first and then select the new window
+    if (*it == originally_selected)
+    {
+        // Edge case: if we have gone full circle around the list back to the original app
+        // then we will wind up in a situation where the original app - now in the second z-order
+        // position - will be swapped with the final app, putting the final app in the second position.
+        for (auto window: tools.info_for(selected).windows())
+            tools.send_tree_to_back(window);
+    }
+    else
+    {
+        tools.swap_tree_order(next_window, active_window);
+    }
+
+    // next_window will be a garbage window in this case, so let's not select it
+    if (it != start_it)
+    {
+        tools.select_active_window(next_window);
+    }
+
+    return *it;
+}
+
+auto ApplicationSelector::complete() -> Application
+{
+    if (!is_active())
+    {
+        mir::log_warning("Cannot call ApplicationSelector::stop when the ApplicationSelector is not active.");
```

Comment:
The logged text is wrong: the function is called `complete()`. Also, this gets spammed to the log when launching apps and is probably not something logging and certainly not a warning the user would care about.